### PR TITLE
Prefabs | Reduce memory consumption on big, deeply nested levels.

### PR DESCRIPTION
**Pull request by AMZN-daimini**
## What does this PR do?

Introduces a mechanism to allow to individually toggle whether a prefab instance should cache its DOM (disabled by default). 
Then hooks it up in the Prefab Focus Handler so that only the focused prefab instance's DOM is cached.

## How was this PR tested?

Tested on a custom level with a deeply nested prefab hierarchy

image
image

Memory consumption went from 6.6GB before the change to 5.4GB after the change. Both measurements taken on a clean Editor loading the level (measurement when creating the level was higher).

Considering the clean Editor before loading the level takes up 2.9GB, it's a 1.2GB difference, about a 32% improvement. Of course memory consumption per entity/prefab is not only impacted by the instance's DOM, so the impact is definitely significant.

Did some manual testing and it doesn't look like this introduces issues or degrades editor performance in a noticeable way, even in debug mode, since most operations are done on the focused prefab instance which is still cached. Operations that include other instances may be slightly slower, so we may want to test that further.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.h [hunk 1]
```diff
@@ -98,6 +98,10 @@ namespace AzToolsFramework
             //! @param context reflection context.
             static void Reflect(AZ::ReflectContext* context);
 
+            //! Whether caching the instance's dom should be enabled by default.
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.h:101]: do you mean "disabled"?

**Review by forkercat — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.h (hunk 1 above)
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.h:101]: Well the variable name is in the positive, so I think the comment makes sense?

Comment (forkercat) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.h:101]: Not sure if I understand it right. What it is doing here is have a static variable that represents the default behavior of whether we want to cache the instance DOM or not. And we don't want it to cache the instance DOM by default. Is it correct?

Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.h:101]: Correct. The comment states that the variable determines if the dom should be cached.

Comment (forkercat) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/Instance.h:101]: okay I see. No problem.

My previous interpretation:
(Whether caching the instance's dom) should be enabled by default.

What it actually indicates:
Whether (caching the instance's dom should be enabled by default).

**Review by lsemp3d — APPROVED** [review]

**Review by nick-l-o3de — APPROVED** [review]

**Pull request merged**